proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,358 @@
+
+	log.Infof("No key rule specified, matching by opcode")
+	return true
+}
+
+func (rule *BinaryMemcacheRule) matchOpcode(oc byte) bool {
+	return bytes.Contains(rule.opCodes, []byte{oc})
+}
+
+func getMemcacheKey(packet []byte, extrasLength, keyLength uint32) string {
+	return string(packet[headerSize+extrasLength : headerSize+(extrasLength)+keyLength])
+}
+
+// L7BinaryMemcacheRuleParser parses protobuf L7 rules to and array of BinaryMemcacheRule
+// May panic
+func L7BinaryMemcacheRuleParser(rule *cilium.PortNetworkPolicyRule) []proxylib.L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		panic(fmt.Errorf("Can't get L7 rules"))
+	}
+	var rules []proxylib.L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br BinaryMemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "opCode":
+				br.opCode = v
+				br.opCodes = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+			case "keyPrefix":
+				br.keyPrefix = v
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				panic(fmt.Errorf("Unsupported key: %s", k))
+			}
+		}
+		if br.opCode == "" {
+			panic(fmt.Errorf("opCode not specified"))
+		}
+		log.Infof("Parsed BinaryMemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// BinaryMemcacheParserFactory implements proxylib.ParserFactory
+type BinaryMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection, injectQueue: make([]queuedInject, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	injectQueue  []queuedInject
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	log.Infof("OnData with offset %d", offset)
+
+	if reply {
+		if p.checkInjectQueue() {
+			return proxylib.INJECT, len(DeniedMsgBase)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	data := (bytes.Join(dataBuffers, []byte{}))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	headerMissing := headerSize - len(data)
```

Comment:
I don't think we can underflow an `int` here.

`headerSize` is a constant positive number, `len()` return type is `int`, so it will be at most maximum `int` value, which substracted from positive number should be over minimum `int` value.